std/algebra/emulated/fields_bw6761/e3.go

```diff
@@ -0,0 +1,412 @@
+	t2 = e.fp.Add(t2, b)
+	t1 := e.fp.Add(c0, c1)
+	tmp := e.fp.Add(d0, d1)
+	t1 = e.fp.Mul(t1, tmp)
+	t1 = e.fp.Sub(t1, a)
+	t1 = e.fp.Sub(t1, b)
+	return &E3{
+		A0: *t0,
+		A1: *t1,
+		A2: *t2,
+	}
+}
+
+// Mul sets z to the *E3-product of x,y, returns z
+func (e Ext3) Mul(x, y *E3) *E3 {
+	// Algorithm 13 from https://eprint.iacr.org/2010/354.pdf
+	t0 := e.fp.Mul(&x.A0, &y.A0)
+	t1 := e.fp.Mul(&x.A1, &y.A1)
+	t2 := e.fp.Mul(&x.A2, &y.A2)
+
+	c0 := e.fp.Add(&x.A1, &x.A2)
+	tmp := e.fp.Add(&y.A1, &y.A2)
+	c0 = e.fp.Mul(c0, tmp)
+	c0 = e.fp.Sub(c0, t1)
+	c0 = e.fp.Sub(c0, t2)
+	c0 = MulByNonResidue(e.fp, c0)
+
+	tmp = e.fp.Add(&x.A0, &x.A2)
+	c2 := e.fp.Add(&y.A0, &y.A2)
+	c2 = e.fp.Mul(c2, tmp)
+	c2 = e.fp.Sub(c2, t0)
+	c2 = e.fp.Sub(c2, t2)
+
+	c1 := e.fp.Add(&x.A0, &x.A1)
+	tmp = e.fp.Add(&y.A0, &y.A1)
+	c1 = e.fp.Mul(c1, tmp)
+	c1 = e.fp.Sub(c1, t0)
+	c1 = e.fp.Sub(c1, t1)
+	t2 = MulByNonResidue(e.fp, t2)
+
+	a0 := e.fp.Add(c0, t0)
+	a1 := e.fp.Add(c1, t2)
+	a2 := e.fp.Add(c2, t1)
+
+	return &E3{
+		A0: *a0,
+		A1: *a1,
+		A2: *a2,
+	}
+}
+
+// Square sets z to the *E3-product of x,x, returns z
+func (e Ext3) Square(x *E3) *E3 {
+
+	// Algorithm 16 from https://eprint.iacr.org/2010/354.pdf
+
+	c6 := e.fp.MulConst(&x.A1, big.NewInt(2))
+	c4 := e.fp.Mul(&x.A0, c6) // x.A0 * xA1 * 2
+	c5 := e.fp.Mul(&x.A2, &x.A2)
+	c1 := MulByNonResidue(e.fp, c5)
+	c1 = e.fp.Add(c1, c4)
+	c2 := e.fp.Sub(c4, c5)
+
+	c3 := e.fp.Mul(&x.A0, &x.A0)
+	c4 = e.fp.Sub(&x.A0, &x.A1)
+	c4 = e.fp.Add(c4, &x.A2)
+	c5 = e.fp.Mul(c6, &x.A2) // x.A1 * xA2 * 2
+	c4 = e.fp.Mul(c4, c4)
+	c0 := MulByNonResidue(e.fp, c5)
+	c4 = e.fp.Add(c4, c5)
+	c4 = e.fp.Sub(c4, c3)
+
+	a0 := e.fp.Add(c0, c3)
+	a1 := c1
+	a2 := e.fp.Add(c2, c4)
+
+	return &E3{
+		A0: *a0,
+		A1: *a1,
+		A2: *a2,
+	}
+}
+
+func (e Ext3) Inverse(x *E3) *E3 {
+	res, err := e.fp.NewHint(inverseE3Hint, 3, &x.A0, &x.A1, &x.A2)
+	if err != nil {
+		// err is non-nil only for invalid number of inputs
+		panic(err)
+	}
+
+	inv := E3{
+		A0: *res[0],
+		A1: *res[1],
+		A2: *res[2],
+	}
+	one := e.One()
+
+	// 1 == inv * x
+	_one := e.Mul(&inv, x)
+	e.AssertIsEqual(one, _one)
+
+	return &inv
+
+}
+
+func (e Ext3) DivUnchecked(x, y *E3) *E3 {
+	res, err := e.fp.NewHint(divE3Hint, 6, &x.A0, &x.A1, &x.A2, &y.A0, &y.A1, &y.A2)
+	if err != nil {
+		// err is non-nil only for invalid number of inputs
+		panic(err)
+	}
+
+	div := E3{
+		A0: *res[0],
+		A1: *res[1],
+		A2: *res[2],
+	}
+
+	// 1 == inv * x
+	_x := e.Mul(&div, x)
+	e.AssertIsEqual(x, _x)
+
+	return &div
+
+}
+
+// MulByNonResidue mul x by (0,1,0)
+func (e Ext3) MulByNonResidue(x *E3) *E3 {
+	z := &E3{
+		A0: x.A2,
+		A1: x.A0,
+		A2: x.A1,
+	}
+	z.A0 = *MulByNonResidue(e.fp, &z.A0)
+	return z
+}
+
+// AssertIsEqual constraint self to be equal to other into the given constraint system
+func (e Ext3) AssertIsEqual(a, b *E3) {
+	e.fp.AssertIsEqual(&a.A0, &b.A0)
+	e.fp.AssertIsEqual(&a.A1, &b.A1)
+	e.fp.AssertIsEqual(&a.A2, &b.A2)
+}
+
+func (e Ext3) Set(x *E3) *E3 {
```

Comment:
Would rename - `Set` makes sense if it is a method on `*E3`. Maybe `Copy` or `Clone`?